.clang-format

```diff
@@ -0,0 +1,27 @@
+BasedOnStyle: LLVM
+IndentWidth: 4
+AllowShortIfStatementsOnASingleLine: false
+AlwaysBreakTemplateDeclarations: Yes
+BinPackArguments: false
+BinPackParameters: false
+BreakBeforeBraces: Attach
+ColumnLimit: 120
+ConstructorInitializerIndentWidth: 4
+Cpp11BracedListStyle: true
+DerivePointerAlignment: false
+IncludeBlocks: Regroup
```

Comment:
I looked into clang-format 15 and it has the following for LLVM:

```
IncludeCategories:
  - Regex:           '^"(llvm|llvm-c|clang|clang-c)/'
    Priority:        2
    CaseSensitive:   false
  - Regex:           '^(<|"(gtest|gmock|isl|json)/)'
    Priority:        3
    CaseSensitive:   false
  - Regex:           '.*'
    Priority:        1
    CaseSensitive:   false
```

which will result in STL headers being at the end of the include list, is this something that is preferred? I would recommend something like (based on LLVM configuration) this in a header or source file:

```
#include <HEADER_OF_CPP_IF_NEEDED>

#include <STL_HEADER_1>
#include <STL_HEADER_2>

#include <OTHER_LIB_HEADER_1>
#include <OTHER_LIB_HEADER_2>

#include <THIRD_PARTY_HEADERS>
```

I would even set this to `Preserved` and let the developer decide, or create a custom `IncludeCategories` block.

Reply:
AFAIK, there is a case where manually organizing headers can be useful: for each .h/.c pairs, e.g., celix_bundle_cache.h(.c), always include the header file (`celix_bundle_cache.h`) in the first line of the non-comment part of the source file (`celix_bundle_cache.c`) to make sure the header file is self-contained, which means it can be used alone.  Self-contained header file is really nice.

Reply:
clang-format does take into account. So when you run formatting over a `foo.c` source file and there is a include of `foo.h` this will be included first. 

This should even be extendable with the `IncludeIsMainRegex` option, but I could not get that to work.

Reply:
I've consistently encountered problemswith this setting while developing libraries, particularly when using angle brackets (`<>`) to include header files within the library. However, these problems don't seem to occur when using double quotes (`""`).

E.g.:

```cpp
// lib/detail/header1.cpp
#include <lib/detail/header1.hpp>

#include <lib/header2.hpp>
#include <lib/header3.hpp>

#include <memory>

#include <spdlog/spdlog.hpp>
```

will get reorganized in:
```cpp
// lib/detail/header1.cpp
#include <memory>

#include <lib/header2.hpp>
#include <lib/header3.hpp>
#include <lib/detail/header1.hpp> // This header should be at the top of the file in my opinion.

#include <spdlog/spdlog.hpp>
```

while, the following:
```cpp
// lib/detail/header1.cpp
#include "lib/detail/header1.hpp"

#include "lib/header2.hpp"
#include "lib/header3.hpp"

#include <memory>

#include <spdlog/spdlog.hpp>
```

will get reorganized in:

```cpp
// lib/detail/header1.cpp
#include "lib/detail/header1.hpp"

#include <memory>

#include "lib/header2.hpp"
#include "lib/header3.hpp"

#include <spdlog/spdlog.hpp>
```

Reply:
I purpose to not regroup includes header for now. 

I will leave this as comments, so that it can be used if desired on case per case basis.
If needed we can revisit this at a later stage.